File: 0206-reverse-linked-list/0206-reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        
//         if(head == null){
//             return head;
//         }
        
//         ListNode previous = null;
//         ListNode current;
        
//         current = head;
        
//         while(current != null){
//             ListNode tmp = current.next;
//             current.next = previous;
//             previous = current;
//             current = tmp;
//         }
        
//         return previous;
        
        if(head == null || head.next == null){
            return head;
        }
        
        ListNode newHead = reverseList(head.next);
        head.next.next = head;
        head.next = null;

        return newHead;
    }
}

File: 0206-reverse-linked-list/0206-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
        if head is None or head.next is None:
            return head
        
        p1, p2 = None, head
        
        while p2.next is not None:
            
            tmp = p2.next
            p2.next = p1
            p1 = p2
            p2 = tmp
            
        p2.next = p1
        
        return p2
            
            